book/src/dev/rfcs/0004-asynchronous-script-verification.md

```diff
@@ -175,23 +337,41 @@ impl PendingUtxos {
 
 The state service should maintain an `Arc<Mutex<PendingUtxos>>`, used as follows:
 
-1. In `Service::call(Request::AwaitUtxo(u))`, the service should:
+1. In `Service::call(Request::AwaitSpendableUtxo { outpoint: u, .. }`, the service should:
   - call `PendingUtxos::queue(u)` to get a future `f` to return to the caller;
-  spawn a task that does a rocksdb lookup for `u`, calling `PendingUtxos::respond(u, output)` if present;
+  - spawn a task that does a rocksdb lookup for `u`, calling `PendingUtxos::respond(u, output)` if present;
   - check the in-memory storage for `u`, calling `PendingUtxos::respond(u, output)` if present;
   - return `f` to the caller (it may already be ready).
-  The common case is that `u` references an old UTXO, so spawning the lookup
+  The common case is that `u` references an old spendable UTXO, so spawning the lookup
   task first means that we don't wait to check in-memory storage for `u`
   before starting the rocksdb lookup.
 
-2. In `Service::call(Request::CommitBlock(block, ..))`, the service should:
-  - call `PendingUtxos::check_block(block.as_ref())`;
+2. In `f`, the future returned by `PendingUtxos::queue(u)`, the service should
+  check that the `TransparentOutputOrigin` is spendable before returning it:
+  - if the `TransparentOutputOrigin` is `NonCoinbase`, return the utxo;
+  - if the `TransparentOutputOrigin` is `Coinbase`, check that:
+    - `spend_pools` is `AllShieldedOutputs`, and
+    - `spend_height` is greater than or equal to
+      `MIN_TRANSPARENT_COINBASE_MATURITY` plus the `Coinbase { height, .. }`,
+    - then return the utxo.
+
+3. In `PendingUtxos::check_against(utxos, height)`, the service should:
+  - return `TransparentOutputOrigin::Coinbase` if `OutPoint.index` is `0`, and
+  - use the block `height` as the `Coinbase { height, .. }`.
```

Comment:
```suggestion
```suggestion
  - return `TransparentOutputOrigin::Coinbase` if `tx_index == 0`,
    using `height` as the coinbase height.
```
```